### PR TITLE
feat: memory model proofs for simp_mem [2/3?] 

### PR DESCRIPTION
**Pull request by bollu**
This PR is stacked on top of https://github.com/leanprover/LNSym/pull/105, and is peeled from https://github.com/leanprover/LNSym/pull/90. This is the second of 3 anticipated PRs which build the new `simp_mem` tactic.

This PR change `mem` from private to public, because it was found when writing automation that 
```lean
theorem ArmState.read_mem_eq_mem_read : read_mem addr s = s.mem.read addr := rfl
```

exposes `s.mem` to the outside world, on which we perform memory operations.
So, we think of `s.mem` as a public interface for the memory, whose *methods* which modify memory will eventually be made private.

### License:

By submitting this pull request, I confirm that my contribution is
made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by alexkeizer — COMMENTED** [review]
Just checking that this works

**Comment by alexkeizer** [comment]
> So, we think of s.mem as a public interface for the memory, whose methods which modify memory will eventually be made private.

This seems dodgy. It feels far more natural to me to have the `Memory` of an `ArmState` be inaccessible, but, if I somehow get my hands on some memory somewhere that I can modify that one

File: Arm/State.lean [hunk 1]
```diff
@@ -785,6 +807,27 @@ theorem getLsb_read_bytes {n i : Nat} {addr : BitVec 64} {m : Memory} (hn : n 
         omega
       · omega
 
+/--
+The describes the behaviour of `m.read_bytes` at a byte level granularity.
```
Comment (alexkeizer) [inline Arm/State.lean:811]:
```suggestion
Describe the behaviour of `m.read_bytes` at a byte level granularity.
```

File: Arm/State.lean [hunk 2]
```diff
@@ -785,6 +807,27 @@ theorem getLsb_read_bytes {n i : Nat} {addr : BitVec 64} {m : Memory} (hn : n 
...
+  rw [getLsb_read_bytes]
+  by_cases h₁ : i * 8 + ↑j < n * 8
+  · simp only [h₁, decide_True, Bool.true_and]
+    simp only [show (i < n) by omega, ↓reduceIte]
+    simp only [show (i * 8 + ↑j) / 8 = i by omega]
+    simp only [show (i * 8 + ↑j) % 8 = j by omega]
+    rfl
+  · simp only [h₁, decide_False, Bool.false_and, Bool.false_eq]
+    simp only [show ¬(i < n) by omega, ↓reduceIte, BitVec.getLsb_zero]
+  · omega
```
Comment (alexkeizer) [inline Arm/State.lean:830]: You don't use `Memory.extractLsByte_read` in this proof, right? 
If not, can we prove `Memory.extractLsByte_read` as a specialization of this theorem with `n = 0`, to get a much shorter proof than we currently have there

**Review by alexkeizer — COMMENTED** [review]
I'm somewhat sceptical of the move to make the `mem` field with the stated intention of making its definitions private. Is it possible to not use the lemma you say causes problems in automation, and instead keep at the more abstract level?

If not, then the change is unavoidable, but it might be worth thinking about that at least a little bit

File: Arm/BitVec.lean [hunk 3]
```diff
@@ -940,14 +940,15 @@ theorem sub_le_sub_iff_right (a b c : BitVec w₁) (hac : c ≤ a)
 
 /--
 Definition to extract the `n`th least significant *Byte* from a bitvector.
-TODO: this should be named `getLsByte`.
+TODO: this should be named `getLsbByte`.
```
[the remark this answers is not in the document]
Comment (bollu) [reply Arm/BitVec.lean]: @shigoel prefers it over `LsByte`, citing concerns of typability.

**Comment by bollu** [comment]
> I'm somewhat sceptical of the move to make the mem field with the stated intention of making its definitions private. Is it possible to not use the lemma you say causes problems in automation, and instead keep at the more abstract level?
> If not, then the change is unavoidable, but it might be worth thinking about that at least a little bit

Unfortunately not, as the simplification hinges on being able to reason on `Memory` directly, without the `ArmState` intervening. I'm not sure I follow why this feels dodgy to you --- I'll encapsulate the behavior of memory with higher level theorems :)

File: Arm/BitVec.lean (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexkeizer) [reply Arm/BitVec.lean]: I'll concede the color of this particular bikeshed

File: Arm/State.lean (hunk 2 above)
Comment (bollu) [reply Arm/State.lean:808]: done! Thanks for the cleanup

File: Arm/State.lean [hunk 4]
```diff
@@ -700,6 +700,28 @@ A variant of `read_mem` that directly talks about writes to memory, instead of o
...
+-/
+@[memory_rules]
+theorem Memory.extractLsByte_read (m : Memory) :
+    (m.read addr).extractLsByte i = if i = 0 then m.read addr else 0#8 := by
+  rw [BitVec.extractLsByte_def]
+  by_cases hi : i = 0
+  · subst hi
+    simp only [bitvec_rules, minimal_theory]
+    apply BitVec.eq_of_getLsb_eq
+    simp
```
[the remark this answers is not in the document]
Comment (bollu) [reply Arm/State.lean]: Done.

**Comment by shigoel** [comment]
> > I'm somewhat sceptical of the move to make the mem field with the stated intention of making its definitions private. Is it possible to not use the lemma you say causes problems in automation, and instead keep at the more abstract level?
> > If not, then the change is unavoidable, but it might be worth thinking about that at least a little bit
> 
> Unfortunately not, as the simplification hinges on being able to reason on `Memory` directly, without the `ArmState` intervening. I'm not sure I follow why this feels dodgy to you --- I'll encapsulate the behavior of memory with higher level theorems :)

The rest of this PR LGTM, thanks, but this discussion is interesting. Thanks, @alexkeizer, for bringing this up!

@bollu Don't your lemmas `State.read_mem_bytes_eq_mem_read_bytes`, `write_mem_bytes_eq_mem_write_bytes`, `ArmState.read_mem_eq_mem_read`, and `ArmState.write_mem_eq_mem_write` help in simplifying away `ArmState` to `Memory`? 

BTW, I notice that `ArmState.write_mem_eq_mem_write` differs from `write_mem_bytes_eq_mem_write_bytes` in that the former talks about the projection of memory from an `ArmState`, whereas the latter talks about `ArmState` only. Why?

Aside: Fix these lemmas to be in the same namespace? Note that `write_mem_bytes_eq_mem_write_bytes` is globally defined. Time to change `section State` in `Arm/State.lean` to a `namespace`?

**Comment by bollu** [comment]
> @bollu Don't your lemmas State.read_mem_bytes_eq_mem_read_bytes, write_mem_bytes_eq_mem_write_bytes, ArmState.read_mem_eq_mem_read, and ArmState.write_mem_eq_mem_write help in simplifying away ArmState to Memory?


@shigoel Indeed, and that's precisely the problem! When they simplify state to memory, the proof state *explicitly* refers to `s.mem`. If at this point, `ArmState.Memory` is private, we're hosed, since we can't rewrite `s.mem`, or even *refer* to `s.mem` (even though such a term shows up in the proof state!) So we must make `s.mem` *itself* private, and hide *its* details.

**Comment by bollu** [comment]
> BTW, I notice that ArmState.write_mem_eq_mem_write differs from write_mem_bytes_eq_mem_write_bytes in that the former talks about the projection of memory from an ArmState, whereas the latter talks about ArmState only. Why?


@shigoel I didn't want to disturb too much of the repo, so I kept around whatever exists, hoping to deprecate the old stuff in subsequent PRs. That seemed far less violent than attempting to both add new features and deprecate old ones simultaneously :)

**Comment by shigoel** [comment]
> @shigoel Indeed, and that's precisely the problem! When they simplify state to memory, the proof state _explicitly_ refers to `s.mem`. If at this point, `ArmState.Memory` is private, we're hosed, since we can't rewrite `s.mem`, or even _refer_ to `s.mem` (even though such a term shows up in the proof state!) So we must make `s.mem` _itself_ private, and hide the details of `Memory` itself, _not_ the field `ArmState.mem`.

Can we define a wrapper around `s.mem`, say `(getMem s)` or something, and use that instead of `s.mem` in these theorems and consequently, proof state? `s.mem` could become private then, right?

**Comment by bollu** [comment]
@shigoel we could, but then I fail to see what this achieves, since `getMem : ArmState -> Memory` is a wrapper around the projection `ArmState.mem : ArmState -> Memory`

**Comment by shigoel** [comment]
From offline discussions:

@shigoel: "It depends on what our public interface for the State is going to be. We seem to be trending in the direction of making all fields of the state private -- the two exceptions right now are memory and program. We could fix the program issue by using `r` to access it (you even opened an issue about it) but this `getMem` could help with the memory field."

@alexkeizer: "To be clear, my original feeling that the suggested setup felt off comes from software engineering sensibilities, I'm not really sure what the best setup with private fields is once proof engineering comes into play"

@bollu: "I would be glad to rethink the API and figure out how to make the field private via a `getMem` / `setMem` pair. It would be nice to do it in another PR, since the blast radius of me trying to do this got larger than I anticipated!"

Merging this PR. Thanks all!

**Review by shigoel — APPROVED** [review]

**Pull request merged**